zcash_client_backend/src/data_api/wallet.rs

```diff
@@ -772,104 +795,108 @@ where
         },
     );
 
+    #[cfg(feature = "transparent-inputs")]
+    let mut has_shielded_inputs = false;
+
     for (sapling_key, sapling_note, merkle_path) in sapling_inputs.into_iter() {
         builder.add_sapling_spend(&sapling_key, sapling_note.clone(), merkle_path)?;
+        #[cfg(feature = "transparent-inputs")]
+        {
+            has_shielded_inputs = true;
+        }
     }
 
     #[cfg(feature = "orchard")]
     for (orchard_note, merkle_path) in orchard_inputs.into_iter() {
         builder.add_orchard_spend(usk.orchard(), *orchard_note, merkle_path.into())?;
+        #[cfg(feature = "transparent-inputs")]
+        {
+            has_shielded_inputs = true;
+        }
     }
 
     #[cfg(feature = "transparent-inputs")]
-    let utxos_spent = {
-        let known_addrs = wallet_db
-            .get_transparent_receivers(account)
-            .map_err(Error::DataSource)?;
+    let mut known_addrs = wallet_db
+        .get_transparent_receivers(account_id)
+        .map_err(Error::DataSource)?;
+    #[cfg(feature = "transparent-inputs")]
+    let mut ephemeral_added = false;
+
+    #[cfg(feature = "transparent-inputs")]
+    let mut metadata_from_address = |addr: TransparentAddress| -> Result<
+        TransparentAddressMetadata,
+        ErrorT<DbT, InputsErrT, FeeRuleT>,
+    > {
+        match known_addrs.get(&addr) {
+            None if !ephemeral_added => {
```

Comment:
I find the mutation in the closure of both `ephemeral_added` and `known_addrs` very difficult to reason about. This closure is doing two things: it's retrieving metadata required to compute the spending key for transparent inputs being spent, and it's also lazily retrieving the ephemeral address metadata and adding it to the set. 

As I think about this, though, I'm skeptical that having `known_addrs` as a local variable makes sense. Suppose a wallet has thousands of addresses? This will be retrieving all of them just to find one, most of the time. It would likely be better to directly request the address metadata for a single address at a time from the database; in most transactions, it'll be exactly one address (either we're shielding funds from a single external taddr, or we're spending funds from a single ephemeral address.)